// checkout-sdk-node/dist/api/access/access.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _errors = require("../../services/errors");
var _http = require("../../services/http");
function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }
function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }
/**
 * Class dealing with the access api
 *
 * @export
 * @class Access
 */
class Access {
  constructor(config) {
    this.config = config;
  }

  /**
   * Request an access token
   *
   * @param {Object} body Access object body.
   * @return {Promise<Object>} A promise to the Access response.
   */
  request(body) {
    var _this = this;
    return _asyncToGenerator(function* () {
      try {
        const response = yield (0, _http.createAccessToken)(_this.config, _this.config.httpClient, body);
        return yield response.json;
      } catch (err) {
        const error = yield (0, _errors.determineError)(err);
        throw error;
      }
    })();
  }

  /**
   * Update the access details in the config.
   *
   * @param {Object} body Access response body.
   * @return {void} 
   */
  updateAccessToken(body) {
    this.config.access = {
      token: body.access_token,
      type: body.token_type,
      scope: body.scope,
      expires: new Date(new Date().getTime() + body.expires_in)
    };
  }
}
exports.default = Access;
