// checkout-sdk-node/dist/api/apm-specific/klarna.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _errors = require("../../services/errors");
var _http = require("../../services/http");
function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }
function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }
/**
 * Class dealing with the /klarna and /klarna-external endpoint
 *
 * @export
 * @class Klarna
 */
class Klarna {
  constructor(config) {
    this.config = config;
  }

  /**
   * Create a session
   *
   * @param {Object} body Sessions details.
   * @return {Promise<Object>} A promise to the Klarna response.
   */
  createSession(body) {
    var _this = this;
    return _asyncToGenerator(function* () {
      const url = _this.config.host.includes('sandbox') ? `${_this.config.host}/klarna-external/credit-sessions` : `${_this.config.host}/klarna/credit-sessions`;
      try {
        const response = yield (0, _http.post)(_this.config.httpClient, url, _this.config, _this.config.sk, body);
        return yield response.json;
      } catch (err) {
        const error = yield (0, _errors.determineError)(err);
        throw error;
      }
    })();
  }

  /**
   * Get a session
   *
   * @param {string} id Session id.
   * @return {Promise<Object>} A promise to the Klarna response.
   */
  getSession(id) {
    var _this2 = this;
    return _asyncToGenerator(function* () {
      const url = _this2.config.host.includes('sandbox') ? `${_this2.config.host}/klarna-external/credit-sessions/${id}` : `${_this2.config.host}/klarna/credit-sessions/${id}`;
      try {
        const response = yield (0, _http.get)(_this2.config.httpClient, url, _this2.config, _this2.config.sk);
        return yield response.json;
      } catch (err) {
        const error = yield (0, _errors.determineError)(err);
        throw error;
      }
    })();
  }

  /**
   * Capture a klarna payment
   *
   * @param {string} id Payment id.
   * @param {Object} body Capture details.
   * @return {Promise<Object>} A promise to the Klarna response.
   */
  capture(id, body) {
    var _this3 = this;
    return _asyncToGenerator(function* () {
      const url = _this3.config.host.includes('sandbox') ? `${_this3.config.host}/klarna-external/orders/${id}/captures` : `${_this3.config.host}/klarna/orders/${id}/captures`;
      try {
        const response = yield (0, _http.post)(_this3.config.httpClient, url, _this3.config, _this3.config.sk, body);
        return yield response.json;
      } catch (err) {
        const error = yield (0, _errors.determineError)(err);
        throw error;
      }
    })();
  }

  /**
   * Void a klarna payment
   *
   * @param {string} id Payment id.
   * @param {Object} [body] Void details.
   * @return {Promise<Object>} A promise to the Klarna response.
   */
  void(id, body) {
    var _this4 = this;
    return _asyncToGenerator(function* () {
      const url = _this4.config.host.includes('sandbox') ? `${_this4.config.host}/klarna-external/orders/${id}/voids` : `${_this4.config.host}/klarna/orders/${id}/voids`;
      try {
        const response = yield (0, _http.post)(_this4.config.httpClient, url, _this4.config, _this4.config.sk, body);
        return yield response.json;
      } catch (err) {
        const error = yield (0, _errors.determineError)(err);
        throw error;
      }
    })();
  }
}
exports.default = Klarna;
